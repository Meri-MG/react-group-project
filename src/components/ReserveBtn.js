import React from 'react';

const ReserveBtn = () => (
  <div>
    <button type="button">Reserve Rocket</button>
  </div>
);

export default ReserveBtn;
