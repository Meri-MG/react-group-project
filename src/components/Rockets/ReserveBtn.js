import React from 'react';

const ReserveBtn = (props) => {
  const rocket = props;
  const { reserve, handleReserve } = rocket;
  return (
    <div>
      {!reserve ? (
        <button
          className="reservationBtn"
          type="button"
          onClick={handleReserve}
        >
          Reserve Rocket
        </button>
      ) : (
        <button
          className="cancelBtn"
          type="button"
          onClick={handleReserve}
        >
          Cancel Reservation
        </button>
      )}
    </div>
  );
};
export default ReserveBtn;
