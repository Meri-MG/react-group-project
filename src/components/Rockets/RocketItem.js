import React from 'react';
import { useDispatch } from 'react-redux';
import ReserveBtn from './ReserveBtn';
import { getReserved } from '../../redux/rockets/rockets';

const RocketItem = (props) => {
  const dispatch = useDispatch();
  const rocket = props;
  const handleReserve = () => {
    const update = !rocket.data.reserve;
    const setReserve = () => ({ btnId: rocket.data.id, reserve: update });
    dispatch(getReserved(setReserve()));
  };

  return (
    <div className="rockets-wrapper">
      <img src={rocket.data.flickr_images[0]} alt="rocket" />
      <div>
        <h2>{rocket.data.rocket_name}</h2>
        <p>
          {rocket.data.reserve && <span className="reserve-badge">Reserved</span>}
          {rocket.data.description}
        </p>
        <ReserveBtn
          reserve={rocket.data.reserve}
          id={rocket.data.id}
          handleReserve={handleReserve}
        />
      </div>
    </div>
  );
};

export default RocketItem;
