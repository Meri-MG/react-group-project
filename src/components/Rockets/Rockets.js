import React, { useEffect } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { useDispatch, useSelector } from 'react-redux';
import { getRocketsFromAPI } from '../../redux/rockets/rockets';
import RocketItem from './RocketItem';

const Rockets = () => {
  const data = useSelector((state) => state.rocketsReducer);
  const dispatch = useDispatch();
  useEffect(() => {
    if (!data[0]) {
      dispatch(getRocketsFromAPI());
    }
  }, []);

  return (
    <div>
      <hr />
      <ul>
        {data.map((rocket) => (
          <RocketItem key={rocket.id} data={rocket} />
        ))}
      </ul>
    </div>
  );
};

export default Rockets;
