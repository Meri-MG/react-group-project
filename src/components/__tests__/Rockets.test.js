import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import * as ReactRedux from 'react-redux';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import store from '../../redux/configureStore';
import Rockets from '../Rockets/Rockets';
import ReserveBtn from '../Rockets/ReserveBtn';

// useDispatch returns a function which we are mocking here
const mockDispatch = jest.fn();

beforeAll(() => {
  ReactRedux.useDispatch = jest.fn().mockImplementation(() => mockDispatch);
});

beforeEach(() => {
  ReactRedux.useDispatch.mockClear();
});

describe('Rockets Component test', () => {
  it('Renders Rocket Page', () => {
    const app = renderer
      .create(
        <Provider store={store}>
          <Rockets />
        </Provider>,
      )
      .toJSON();
    expect(app).toMatchSnapshot();
  });
  it('test click ', async () => {
    const rocket = {
      data: {
        reserve: false,
        id: 'item1',
      },
    };
    const handleReserve = jest.fn();
    render(
      <ReserveBtn
        reserve={rocket.data.reserve}
        id={rocket.data.id}
        handleReserve={handleReserve}
      />,
    );
    await userEvent.click(screen.getByRole('button'));
    expect(handleReserve).toHaveBeenCalled();
  });
  it('Button Text', async () => {
    const rocket = {
      data: {
        reserve: true,
        id: 'item1',
      },
    };
    const handleReserve = jest.fn();
    render(
      <ReserveBtn
        reserve={rocket.data.reserve}
        id={rocket.data.id}
        handleReserve={handleReserve}
      />,
    );
    await userEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('button').textContent).toEqual('Cancel Reservation');
  });
});
