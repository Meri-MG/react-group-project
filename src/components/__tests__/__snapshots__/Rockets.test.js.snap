// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Component test Renders Rocket Page 1`] = `
<div>
  <hr />
  <ul />
</div>
`;
